### PR TITLE
Add IVsCredentialProvider, load it via MEF

### PR DESCRIPTION
**Pull request by pspill**

### CODE REVIEW

File: src/NuGet.Clients/NuGet.Credentials/CredentialService.cs [hunk 1]
```diff
@@ -35,12 +35,12 @@ public class CredentialService : ICredentialService
...
         /// Constructor
         /// </summary>
-        /// <param name="providers">Enumeration of credential providers.</param>
+        /// <param name="providersDelegate">Used to enumerate available credential providers.</param>
         /// <param name="errorDelegate">Used to write error messages to the user</param>
         /// <param name="nonInteractive">If true, the nonInteractive flag will be passed to providers.
         /// NonInteractive requests must not promt the user for credentials.</param>
         public CredentialService(
-            IEnumerable<ICredentialProvider> providers,
+            Func<IEnumerable<ICredentialProvider>> providersDelegate,
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Credentials/CredentialService.cs]: Null-check please.

File: src/NuGet.Clients/NuGet.Credentials/CredentialService.cs [hunk 2]
```diff
@@ -51,13 +51,13 @@ public class CredentialService : ICredentialService
...
             _nonInteractive = nonInteractive;
-            Providers = providers ?? new List<ICredentialProvider>();
+            Providers = new Lazy<IEnumerable<ICredentialProvider>>(providersDelegate);
         }
 
         /// <summary>
         /// Gets the currently configured providers.
         /// </summary>
-        public IEnumerable<ICredentialProvider> Providers { get; } = null;
+        public Lazy<IEnumerable<ICredentialProvider>> Providers { get; } = null;
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Credentials/CredentialService.cs]: default null is not necessary

Comment (alpaix) [inline src/NuGet.Clients/NuGet.Credentials/CredentialService.cs]: To maintain the contract I'd keep the original property and would add a backup field, e.g. `_providersFactory` instead.

File: src/NuGet.Clients/NuGet.Credentials/CredentialStatus.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿namespace NuGet.Credentials
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Credentials/CredentialStatus.cs]: Copyright header please.

File: src/NuGet.Clients/VisualStudio/Extensibility/IVsCredentialProvider.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        /// null should be returned.</param>
+        /// <param name="isRetry">True if credentials were previously acquired for this uri, but
+        /// the supplied credentials did not allow authorized access.</param>
+        /// <param name="nonInteractive">If true, then interactive prompts must not be allowed.</param>
+        /// <param name="cancellationToken">This cancellation token should be checked to determine if the
+        /// operation requesting credentials has been cancelled.</param>
+        /// <returns>Credentials acquired by this provider for the given package source uri.
+        /// If the provider does not handle requests for the input parameter set, then null should be returned.
+        /// If the provider does handle the request, but cannot supply credentials, an exception should be thrown.</returns>
+        Task<ICredentials> Get(Uri uri,
```
Comment (alpaix) [inline src/NuGet.Clients/VisualStudio/Extensibility/IVsCredentialProvider.cs]: Can we rename it to `GetCredentialsAsync` or at least to `GetAsync` to be consistent on naming convention?

File: test/NuGet.Clients.Tests/NuGet.VsExtension.Test/AssertExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace NuGet.VsExtension.Test
+{
+    public static class AssertExtensions
+    {
+        public static async Task<T> RecordExceptionAsync<T>(Func<Task> task) where T : Exception
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.VsExtension.Test/AssertExtensions.cs]: Use `Assert.ThrowsAsync<ArgumentException>` instead.

File: src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
```
Comment (alpaix) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs]: header

File: src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
+﻿using System;
```
Comment (alpaix) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs]: header here

File: src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        /// Constructor, allows changing initializer Action for testing purposes
+        /// </summary>
+        /// <param name="dte">DTE instance, used to determine the Visual Studio version.</param>
+        /// <param name="fallbackProviderFactory">Factory method used to create a fallback provider for
+        /// Dev14 in case a VSTS credential provider can not be imported.</param>
+        /// <param name="errorDelegate">Used to write error messages to the user.</param>
+        /// <param name="initializer">Init method used to supply MEF imports. Should only
+        /// be supplied by tests.</param>
+        public VsCredentialProviderImporter (
+            EnvDTE.DTE dte,
```
Comment (alpaix) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs]: null-check for mandatory parameters

File: src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+{
+    /// <summary>
+    /// Wraps an IVsCredentialProvider.  IVsCredentialProvider ensures that VS Extensions 
+    /// can supply credential providers implementing a stable interface across versions.
+    /// </summary>
+    public class VsCredentialProviderAdapter : ICredentialProvider
+    {
+        private readonly IVsCredentialProvider _provider;
+
+        public VsCredentialProviderAdapter(IVsCredentialProvider provider)
```
Comment (alpaix) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs]: null-check

File: src/NuGet.Clients/NuGet.Credentials/CredentialService.cs (hunk 1 above)
Comment (pspill) [inline src/NuGet.Clients/NuGet.Credentials/CredentialService.cs]: fixed

File: src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs (hunk 8 above)
Comment (pspill) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderImporter.cs]: fixed

File: src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs (hunk 9 above)
Comment (pspill) [inline src/NuGet.Clients/VsExtension/VsCredentialProviderAdapter.cs]: fixed

**Comment by alpaix** [comment]
Looks good! Thanks!
:shipit: 

Consider updating [documentation](https://docs.nuget.org/create/extensibility-apis) and [samples](https://github.com/NuGet/samples).

File: src/NuGet.Clients/VsExtension/NuGetPackage.cs [hunk 10]
```diff
@@ -329,23 +329,9 @@ protected override void Initialize()
...
-                    NuGet.NullCredentialProvider.Instance, packageSourceProvider)),
-                new VisualStudioAccountProvider(),
-                new VisualStudioCredentialProvider(webProxy)
-            };
-
             var credentialService = new CredentialService(
-                credentialProviders,
-                (s) => this._outputConsoleLogger.OutputConsole.WriteLine(s),
+                GetCredentialProviders,
+                this._outputConsoleLogger.OutputConsole.WriteLine,
```
Comment (yishaigalatzer) [inline src/NuGet.Clients/VsExtension/NuGetPackage.cs]: don't use this.

**Pull request closed without merging**